blockchain/chain_sync/src/sync.rs

```diff
@@ -478,13 +461,53 @@ where
         }
         Ok(())
     }
+    /// Syncs forked chains
+    async fn sync_fork(&mut self, head: &Tipset, to: &Tipset) -> Result<Vec<Tipset>, Error> {
+        // TODO change from using random peerID to managed
+        let peer_id = PeerId::random();
+        // pulled from Lotus: https://github.com/filecoin-project/lotus/blob/master/chain/sync.go#L996
+        const FORK_LENGTH_THRESHOLD: u64 = 500;
+
+        // Load blocks from network using blocksync
+        let tips: Vec<Tipset> = match self
+            .network
+            .blocksync_headers(peer_id.clone(), head.parents(), FORK_LENGTH_THRESHOLD)
+            .await
+        {
+            Ok(ts) => ts,
+            Err(_e) => {
+                return Err(Error::Other("Failed to retrieve tipset".to_string()));
+            }
+        };
 
-    async fn sync_fork(&mut self, _head: &Tipset, _to: &Tipset) -> Result<Vec<Arc<Tipset>>, Error> {
-        // TODO sync fork until tipsets are equal or reaches genesis
-        todo!()
+        let mut ts = self.chain_store.tipset_from_keys(to.parents())?;
+
+        for i in 0..tips.len() {
+            if *ts.tip_epoch().chain_epoch() == 0 {
+                if let Some(gen) = self.chain_store.genesis()? {
+                    if gen != ts.blocks()[0] {
+                        return Err(Error::Other(
+                            "Chain is linked back to a different genesis (bad genesis)".to_string(),
+                        ));
+                    }
+                }
+                return Err(Error::Other(
+                    "Synced chain forked at genesis, refusing to sync".to_string(),
+                ));
+            }
+            if ts.equals(tips[i].clone()) {
```

Comment:
Shouldn't need to clone to check equality, if you do decide to keep that function, you should check equality with a ref to the other tipset